### PR TITLE
feat: Add spec.AdditionalArgs to ThanosRuler

### LINKED ISSUE #5260: Add spec.additionalArgs to ThanosRuler CRD
**What is missing?**

A way to pass additional arguments to the ThanosRuler CRD as we have for Prometheus CRD

**Why do we need it?**

Different environments have different needs, we should allow prom-operator users to be able to run ThanosRuler with arguments other than the ones set by the operator

**Environment**

* Prometheus Operator version:

    `Insert image tag or Git SHA here`
    
    

**Anything else we need to know?**:

### PR DESCRIPTION
**Pull request by xiu**
Fixes #5260

## Description

Adds `spec.AdditionalArgs` to ThanosRuler

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Add `spec.AdditionalArgs` to ThanosRuler
```

### CODE REVIEW

**Comment by xiu** [comment]
I'm not sure why `TestOperatorUpgrade` is failing, I've successfully executed it in minikube. Let me know if you have any more insights into it.

File: pkg/thanos/statefulset.go [hunk 1]
```diff
@@ -530,3 +536,62 @@ func mountSecret(secretSelector *v1.SecretKeySelector, volumeName string, trVolu
...
+		if arg.Value != "" {
+			containerArgs = append(containerArgs, fmt.Sprintf("--%s=%s", arg.Name, arg.Value))
+		} else {
+			containerArgs = append(containerArgs, fmt.Sprintf("--%s", arg.Name))
+
+		}
+	}
+
+	return containerArgs, nil
+}
```
Comment (xiu) [inline pkg/thanos/statefulset.go:597]: This is a copy of the same functions from `pkg/prometheus/statefulset.go`. I'd rather have them elsewhere than copy them around, suggestions welcome :)

Comment (slashpai) [reply pkg/thanos/statefulset.go:597]: I think `pkg/operator` is better place for common functions and methods

Comment (xiu) [reply pkg/thanos/statefulset.go]: Thanks @slashpai! Moved it there 👍

File: pkg/thanos/statefulset.go [hunk 2]
```diff
@@ -220,78 +220,84 @@ func makeStatefulSetSpec(tr *monitoringv1.ThanosRuler, config Config, ruleConfig
...
 	if tr.Spec.AlertQueryURL != "" {
-		trCLIArgs = append(trCLIArgs, fmt.Sprintf("--alert.query-url=%s", tr.Spec.AlertQueryURL))
+		trCLIArgs = append(trCLIArgs, monitoringv1.Argument{Name: "alert.query-url", Value: tr.Spec.AlertQueryURL})
+	}
+
+	containerArgs, err := operator.BuildArgs(trCLIArgs, tr.Spec.AdditionalArgs)
+	if err != nil {
+		return nil, err
 	}
+	containerArgs = append([]string{"rule"}, containerArgs...)
```
Comment (JoaoBraveCoding) [inline pkg/thanos/statefulset.go:300]: nit: Maybe adding a small comment here would be nice for people not familiar with ThanosRuler config that has to start with `rule`

**Review by JoaoBraveCoding — APPROVED** [review]
Small nit but apart from that LGTM

File: pkg/thanos/statefulset.go (hunk 2 above)
Comment (xiu) [reply pkg/thanos/statefulset.go:302]: Good point, added it 👍

**Review by simonpasquier — APPROVED** [review]
nice!

**Pull request merged**